presto-hive/src/main/java/com/facebook/presto/hive/HiveFileRenamer.java

```diff
@@ -0,0 +1,115 @@
+
+import com.facebook.presto.spi.ConnectorMetadataUpdateHandle;
+import com.facebook.presto.spi.QueryId;
+import com.facebook.presto.spi.SchemaTableName;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class HiveFileRenamer
+{
+    private final ConcurrentMap<QueryId, Map<HiveMetadataUpdateKey, AtomicLong>> queryPartitionFileCounterMap = new ConcurrentHashMap<>();
+    private final ConcurrentMap<QueryId, Map<HiveMetadataUpdateHandle, String>> queryHiveMetadataResultMap = new ConcurrentHashMap<>();
+
+    public List<ConnectorMetadataUpdateHandle> getMetadataUpdateResults(List<ConnectorMetadataUpdateHandle> metadataUpdateRequests, QueryId queryId)
+    {
+        List<ConnectorMetadataUpdateHandle> metadataUpdateResults = new ArrayList<>();
+
+        for (ConnectorMetadataUpdateHandle connectorMetadataUpdateHandle : metadataUpdateRequests) {
+            HiveMetadataUpdateHandle request = (HiveMetadataUpdateHandle) connectorMetadataUpdateHandle;
+            String fileName = getFileName(request, queryId);
+            metadataUpdateResults.add(new HiveMetadataUpdateHandle(request.getRequestId(), request.getSchemaTableName(), request.getPartitionName(), Optional.of(fileName)));
+        }
+        return metadataUpdateResults;
+    }
+
+    public void cleanup(QueryId queryId)
+    {
+        queryPartitionFileCounterMap.remove(queryId);
+        queryHiveMetadataResultMap.remove(queryId);
+    }
+
+    private String getFileName(HiveMetadataUpdateHandle request, QueryId queryId)
+    {
+        if (!queryPartitionFileCounterMap.containsKey(queryId) || !queryHiveMetadataResultMap.containsKey(queryId)) {
+            queryPartitionFileCounterMap.putIfAbsent(queryId, new ConcurrentHashMap<>());
+            queryHiveMetadataResultMap.putIfAbsent(queryId, new ConcurrentHashMap<>());
+        }
+
+        // To keep track of the file counter per query per partition
+        Map<HiveMetadataUpdateKey, AtomicLong> partitionFileCounterMap = queryPartitionFileCounterMap.get(queryId);
+
+        // To keep track of the file name result per query per request
+        // This is to make sure that request - fileName mapping is 1:1
+        Map<HiveMetadataUpdateHandle, String> hiveMetadataResultMap = queryHiveMetadataResultMap.get(queryId);
+
+        // If we have seen this request before then directly return the result.
+        if (hiveMetadataResultMap.containsKey(request)) {
+            // We come here if for some reason the worker did not receive the fileName and it retried the request.
+            return hiveMetadataResultMap.get(request);
+        }
+
+        HiveMetadataUpdateKey key = new HiveMetadataUpdateKey(request);
+        partitionFileCounterMap.putIfAbsent(key, new AtomicLong());
```

Comment:
Do we wanna start with `new AtomicLong(1)`? Actually, check the bucketed case. If bucketed file name started with `1` then we can use `AtomicLong(1)`; otherwise `AtomicLong(0)`.

Reply:
In bucketed case the file name starts with 0. `AtomicLong()` is same as `AtomicLong(0)` . But will change it to `AtomicLong(0)` as suggested, to be on safe side.